enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,764 @@
+The downside of it is that these won't be able to add more racks, but provides a nice way of scaling for static topology clusters.
+
+Datacenter level `nodes` cannot be used together with `nodes` specified per rack. Webhook will prevent it.
+
+The balancing algorithm is iterative, adding/removing one node at a time in each iteration to ensure that racks stay as balanced as possible.
+
+```go
+func balanceNodesAcrossRacks(racks []*appsv1.StatefulSet, totalNodes int) []*appsv1.StatefulSet {
+    // Calculate the total number of existing nodes
+    existingNodes := 0
+    for _, rack := range racks {
+       existingNodes += rack.Replicas
+    }
+
+    if existingNodes == totalNodes {
+       // The desired number of nodes reached, do nothing.
+       return racks
+    }
+    
+    if existingNodes < totalNodes {
+       // Add a node to the least populated rack
+       minRackIndex := findSmallestStatefulSet(racks)
+       racks[minRackIndex].Replicas++
+       return balanceNodesAcrossRacks(racks, totalNodes)
+    }
+    
+    // Remove a node from biggest rack.
+    maxRackIndex := findBiggestStatefulSet(racks)
+    racks[maxRackIndex].Replicas--
+    return balanceNodesAcrossRacks(racks, totalNodes)
+}
+```
+
+New Condition will be added to the `status` of `v1alpha1.ScyllaDBDatacenter`, providing a simple way of noticing rack imbalance.
+Its type is going to be `UnbalancedRacks` and the value of it will be reconciled accordingly to the state of rack imbalance.
+
+--- 
+
+An example of nine node cluster having three racks with uniform sizes and resources.
+```yaml
+spec:
+  scylladb:
+    image: docker.io/scylladb/scylla:6.0.1
+    resources:
+      requests:
+        cpu: 1
+        memory: 4Gi
+    storage:
+      capacity: 100Gi
+      storageClassName: scylladb-local-xfs
+  scyllaDBManagerAgent:
+    image: docker.io/scylladb/scylla-manager-agent:3.3.0
+    resources:
+      requests:
+        cpu: 100m
+        memory: 100Mi
+  topologyLabelSelector:
+    topology.kubernetes.io/region: us-east1
+  nodes: 9
+  racks:
+  - name: a
+    topologyLabelSelector:
+      topology.kubernetes.io/zone: us-east1a
+  - name: b
+    topologyLabelSelector:
+      topology.kubernetes.io/zone: us-east1b
+  - name: c
+    topologyLabelSelector:
+      topology.kubernetes.io/zone: us-east1c
+
+```
+
+### API Spec
+
+Proposed API spec:
+```go
+
+const (
+	AvailableCondition   = "Available"
+	ProgressingCondition = "Progressing"
+	DegradedCondition    = "Degraded"
+)
+
+// ScyllaDBDatacenterSpec defines the desired state of ScyllaDBDatacenter.
+type ScyllaDBDatacenterSpec struct {
+	// metadata controls shared metadata for all pods created based on this spec.
+	// +optional
+	Metadata ObjectTemplateMetadata `json:"metadata,omitempty"`
+
+	// clusterName specifies the name of the scylla cluster.
+	// When joining two DCs, their cluster name must match.
+	ClusterName string `json:"clusterName"`
```

Comment:
please describe where the default comes from

Reply:
optional+omitEmpty

Reply:
As discussed on the meeting, I was afraid it's immutable and indeed it is.
Scylla refuses to boot when cluster name is changed:
```
ERROR 2024-07-11 15:37:31,392 [shard 0:main] init - Startup failed: exceptions::configuration_exception (Saved cluster name a2x2j8q69 != configured name a6hxxm6c8)
```

So this has to be required.

Reply:
I agree, we shouldn't let users be surprised about a default they can't change later.